File: Java/Problem2.java
import java.util.HashMap;

class LRUCache {
    class Node{
        int key, value;
        Node prev, next;

        public Node(int key, int value){
            this.key = key;
            this.value = value;
        }
    }

    private int capacity;
    private HashMap<Integer, Node> map;
    private Node head, tail;
    
    
    public LRUCache(int capacity) {
        this.capacity = capacity;    
        this.map = new HashMap<>();

        
        head = new Node(0,0);
        tail = new Node(0,0);
        head.next = tail;
        tail.prev = head;
    }


    public void remove(Node node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }


    public void insertFirst(Node node){
        node.next = head.next;
        node.prev = head;

        head.next.prev = node;
        head.next = node;
    }

    
    public int get(int key) {
        int value = -1;
        
        if(map.containsKey(key)){
            Node node = map.get(key);
            remove(node);
            insertFirst(node);

            value = node.value;
        }

        return value;
    }


    public void put(int key, int value) {
        if (map.containsKey(key)){
            remove(map.get(key));
        }

        Node first = new Node(key, value);
        insertFirst(first);
        map.put(key, first);

        if (map.size() > capacity){
            Node last = tail.prev;
            remove(last);
            map.remove(last.key);
        }
    }
}

File: Python/Problem2.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.map = {}

        self.head = Node(0,0)
        self.tail = Node(0,0)

        self.head.next = self.tail
        self.tail.prev = self.head

    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def insert_first(self, node):
        node.next = self.head.next
        node.prev = self.head

        self.head.next.prev = node
        self.head.next = node

    def get(self, key: int) -> int:
        value = -1

        if key in self.map:
            node = self.map[key]
            self.remove(node)
            self.insert_first(node)
            value = node.value

        return value


    def put(self, key: int, value: int) -> None:
        if key in self.map:
            self.remove(self.map[key])
        
        first = Node(key, value)
        self.insert_first(first)
        self.map[key] = first

        if len(self.map) > self.capacity:
            last = self.tail.prev
            self.remove(last)
            del self.map[last.key]